logic/atrium-logic/src/commonMain/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt

```diff
@@ -135,9 +138,25 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
 
         val assertions = ArrayList<Assertion>(2)
         assertions.add(createExplanatoryAssertionGroup(container, assertionCreatorOrNull))
-        val mismatches = createIndexAssertions(list) { (_, element) ->
-            mismatchIf(allCreatedAssertionsHold(container, element, assertionCreatorOrNull))
+        val mismatches = list.mapIndexedNotNull { index, e ->
+            val assertion = if (assertionCreatorOrNull != null) {
+                container.collectBasedOnSubject(Some(e as E), assertionCreatorOrNull)
+            } else {
+                // TODO: assertionCreatorOrNull should be non-nullable, and this "to be null" should be a part of expectation
+                assertionBuilder.createDescriptive(DescriptionBasic.TO_BE, Text.NULL) {
+                    e == null
+                }
```

Comment:
It is strange to handle `e == null` here. I wish `assertionCreatorOrNull` was non-nullable.

Reply:
I agree, this functionality could be refactored. Is due to some legacy behaviour introduced in Atrium 0.2.0 or the like and wouldn't be necessary any more. Going to create an issue for it